src/core/audio/AudioAlsa.cpp

```diff
@@ -139,6 +142,59 @@ QString AudioAlsa::probeDevice()
 
 
 
+/**
+ * @brief Creates a list of all available devices.
+ *
+ * Uses the hints API of ALSA to collect all devices. This also includes plug
+ * devices. The reason to collect these and not the raw hardware devices
+ * (e.g. hw:0,0) is that hardware devices often have a very limited number of
+ * supported formats, etc. Plugs on the other hand are software components that
+ * map all types of formats and inputs to the hardware and therefore they are
+ * much more flexible and more what we want.
+ *
+ * Further helpful info http://jan.newmarch.name/LinuxSound/Sampled/Alsa/.
+ *
+ * @return A collection of devices found on the system.
+ */
+AudioAlsa::DeviceInfoCollection AudioAlsa::getAvailableDevices()
+{
+	DeviceInfoCollection deviceInfos;
+
+	char **hints;
+
+	/* Enumerate sound devices */
+	int err = snd_device_name_hint(-1, "pcm", (void***)&hints);
+	if (err != 0)
+	{
+		return deviceInfos;
+	}
+
+	char** n = hints;
+	while (*n != NULL)
+	{
+		char *name = snd_device_name_get_hint(*n, "NAME");
+		char *description = snd_device_name_get_hint(*n, "DESC");
+
+		if (name != 0 && description != 0)
+		{
+			deviceInfos.push_back(DeviceInfo(QString(name), QString(description)));
+		}
+
+		free(name);
+		free(description);
+
```

Comment:
Yes, that was the intention. That's also the reason why sometimes you can see code like this:
```
delete m_pointer;
m_pointer = 0;
```
It's ok to call `delete` on a null pointer but not ok to delete the same pointer (address) twice. By setting it to `0` after the deletion you make sure that you will only run into the valid case in case the code is called again without allocating something new for `m_pointer`.